envoy/server/factory_context.h

```diff
@@ -264,12 +269,44 @@ class FactoryContext : public virtual CommonFactoryContext {
   virtual ProcessContextOptRef processContext() PURE;
 };
 
+/**
+ * Context for downstream network and HTTP filters.
+ */
+class FilterFactoryContext {
+public:
+  FilterFactoryContext(ServerFactoryContext& server_context,
+                       OptRef<DownstreamFactoryContext> downstream_context)
+      : server_context_(server_context), downstream_context_(downstream_context) {}
+
+  /**
+   * @return ServerFactoryContext which lifetime is no shorter than the server.
+   */
+  ServerFactoryContext& getServerFactoryContext() { return server_context_; }
+
+  OptRef<DownstreamFactoryContext> getDownstreamFactoryContext() { return downstream_context_; }
```

Comment:
@kyessenov downstreamFactoryContext would be available for downstream filters, not for upstream filters.  Lifetime would be lifetime of the downstream listener